### PR TITLE
Make Observable.fromInputStream respect the chunkSize more #1154

### PR DESCRIPTION
**Pull request by fnqista**
Fix Observable.fromInputStream not filling up the buffer according to the chunk size passed in, in cases where:
- InputStream actually has more bytes to stream than the chunkSize
- InputStream.read results in bytes streamed less than the chunkSize

### CODE REVIEW

**Comment by fnqista** [comment]
`monix.reactive.internal.builders.CharsReaderObservableSuite` failed under the ci-js job for scala 2.11 and 2.12 cross-build. Anyone knows why? I did not touch this part.

**Comment by Avasil** [comment]
> `monix.reactive.internal.builders.CharsReaderObservableSuite` test `fromCharsReader closes the file handle on cancel` failed under the ci-js job for scala 2.11 and 2.12 cross-build, but was ok for 2.13. Anyone knows why? I did not touch this part.

It's flaky test, I restarted the build

**Comment by fnqista** [comment]
@Avasil It might be an omen that the previous test failed - because when I had a look at it, it seems it will suffer the same fate as issue #1154?

```Reader.read``` falls under the same contract as ```InputStream.read``` more or less. I think it should also be fixed.

**Comment by fnqista** [comment]
Added some more fixes:

Fixed CharsReaderObservable similarly
Added more edge cases for InputStreamObservable:
  - when byte size is 0
  - when provided chunk size is 0
Added more edge cases for CharsReaderObservable:
  - when string is empty
  - when provided chunk size is 0

**Comment by Avasil** [comment]
Awesome, it will be great to finally fix that flaky test :D 
Sorry for delay, I will review today after work!

**Comment by fnqista** [comment]
@Avasil sorry man, I did not really fix the issue with the flimsy test - I only fixed everything related to the chunk sizing as mentioned in #1154 , for both `Observable.fromInputStream` and `Observable.fromCharsReader`, which accepts a `chunkSize` and cannot fill it up in some cases.

This time that flimsy `monix.reactive.internal.builders.CharsReaderObservableSuite` test `fromCharsReader closes the file handle on cancel` did not fail - sorry but it may be by chance.

File: monix-reactive/shared/src/main/scala/monix/reactive/internal/builders/InputStreamObservable.scala [hunk 1]
```diff
@@ -38,10 +38,12 @@ private[reactive] final class InputStreamObservable(in: InputStream, chunkSize:
   self =>
 
   private[this] val wasSubscribed = Atomic(false)
+  private[this] val fallbackChunkSize = 4096
```
[the remark this answers is not in the document]
Comment (fnqista) [reply monix-reactive/shared/src/main/scala/monix/reactive/internal/builders/InputStreamObservable.scala:41]: Yeah, it hangs, since the read will always be zero, the process will be stuck in `fastLoop`, so we should somehow avoid a zero-sized buffer.

The `fallbackChunkSize` I used is the same as what the default value in `Observable.fromInputStream` or `Observable.fromCharsReader`. 

I like the idea of signaling an error instead, because it is a bad input anyway. Let us hear what the others have to say.

[the remark this answers is not in the document]
Comment (alexandru) [reply monix-reactive/shared/src/main/scala/monix/reactive/internal/builders/InputStreamObservable.scala:41]: 👍 signal error.

File: monix-reactive/shared/src/main/scala/monix/reactive/Observable.scala [hunk 2]
```diff
@@ -5010,8 +5010,8 @@ object Observable extends ObservableDeprecatedBuilders {
     *  - ensures that the input stream is closed on completion,
     *    failure or cancellation
     *
-    * @param in is the `Task[InputStream]` generator to convert into an observable
-    * @param chunkSize is the maximum length of the emitted arrays of bytes
+    * @param in the `Task[InputStream]` generator to convert into an observable
+    * @param chunkSize the maximum length of the emitted arrays of bytes, must be positive
```
Comment (fnqista) [inline monix-reactive/shared/src/main/scala/monix/reactive/Observable.scala:5014]: I patterned the chunkSize restriction scaladoc to Observable.range, where it didn't really specify that an IllegalArgumentException can be thrown, but just mentioned in the scaladoc what is valid.

File: monix-reactive/shared/src/main/scala/monix/reactive/internal/builders/InputStreamObservable.scala [hunk 3]
```diff
@@ -35,7 +35,8 @@ import scala.util.control.NonFatal
 import scala.util.{Failure, Success}
 
 private[reactive] final class InputStreamObservable(in: InputStream, chunkSize: Int) extends Observable[Array[Byte]] {
-  self =>
+
+  require(chunkSize > 0, "chunkSize > 0")
```
Comment (fnqista) [inline monix-reactive/shared/src/main/scala/monix/reactive/internal/builders/InputStreamObservable.scala:39]: Again, patterned this fix after `RangeObservable`'s `require(step != 0, "step != 0")`

File: monix-reactive/shared/src/test/scala/monix/reactive/internal/builders/CharsReaderObservableSuite.scala [hunk 4]
```diff
@@ -56,6 +58,20 @@ object CharsReaderObservableSuite extends SimpleTestSuite {
     assert(s.state.tasks.isEmpty, "should be left with no pending tasks")
   }
 
+  test("fromCharsReaderUnsafe should throw if the chunkSize is zero or negative") {
```
Comment (fnqista) [inline monix-reactive/shared/src/test/scala/monix/reactive/internal/builders/CharsReaderObservableSuite.scala:61]: I created two similar (kind-of) tests, one for Unsafe and one for safe. This is because calling `fromCharsReaderUnsafe` throws the error directly, while using `fromCharsReader` will signal via `onError`. If you think the test is redundant, please advise which one to keep.

File: monix-reactive/shared/src/main/scala/monix/reactive/internal/builders/InputStreamObservable.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (fnqista) [reply monix-reactive/shared/src/main/scala/monix/reactive/internal/builders/InputStreamObservable.scala]: Thanks for the advice @alexandru. @Avasil , I updated the PR based on the advice, and I wrote some comments on the file changes to explain some implementation decisions.

**Comment by fnqista** [comment]
The test doesn't fail locally, but fails in travis with 

```[info] - fromInputStreamUnsafe should throw if the chunkSize is zero or negative *** FAILED ***

[info]   Problem with test spec, expecting `Unit`, but received: java.lang.IllegalArgumentException: requirement failed: chunkSize > 0  (InputStreamObservableSuite.scala:67)```

**Comment by alexandru** [comment]
I'd like to download and test it on my machine.

Will do so tonight. If I don't manage to do that today, then please proceed to merge this at your convenience.

**Comment by fnqista** [comment]
Thanks in advance @alexandru  and @Avasil . Regardless of what happens, I would like to applaud you on how wonderful the `monix` library is, it was really a pleasure working with it coming from RxScala/RxJava, it is certainly easier to use and more polished.

I wanted to try it for a long time ever since I saw Alex's talk in Scaladays Copenhagen 2017 and I'm glad I finally got the chance.

**Comment by fnqista** [comment]
Resolved conflicts with #1184 (too fine-grained blocking)
